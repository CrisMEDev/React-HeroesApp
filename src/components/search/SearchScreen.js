import React, { useMemo } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';


export const SearchScreen = ({ history }) => {

    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const initialForm = { searchHero: q };
    const [ formValuesState, handleInputChange, reset ] = useForm( initialForm );

    const { searchHero } = formValuesState;

    // Use memo usado para que no se renderice con cada caracter tecleado
    const heroesFiltered = useMemo(() => getHeroesByName( q ), [ q ]);
    // const heroesFiltered = getHeroesByName( searchHero );

    const handleSearch = (event) => {
        event.preventDefault();

        if ( searchHero.trim().length <= 3 ) return;

        history.push(`?q=${ searchHero }`);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            name="searchHero"
                            value={searchHero}
                            onChange={handleInputChange}
                            autoComplete="off"
                        />

                        <button
                            type="submit"
                            className="btn m-1 w-100 btn-outline-primary"
                        >
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    {
                        ( q === '' )
                        &&  <div className="alert alert-info">
                                Search a hero
                            </div>
                    }

                    {
                        ( q !== '' && heroesFiltered.length === 0 )
                        &&  <div className="alert alert-danger">
                                There is no a hero with { q }
                            </div>
                    }


                    {
                        heroesFiltered.map( (hero) => (
                            <HeroCard
                                key={ hero.id }
                                { ...hero }
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
