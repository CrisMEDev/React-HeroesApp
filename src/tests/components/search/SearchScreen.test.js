import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';

import { SearchScreen } from '../../../components/search/SearchScreen';


describe('Pruebas en <SearchScreen />', () => {
    
    test('Debe de correctamente con valores por defecto', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Search a hero');
        
    });
    
    test('Debe de mostrar a batman y el input con el valor de queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('batman');
        expect( wrapper ).toMatchSnapshot();
        
    });

    test('Debe de mostrar un mensaje de error si no se muestra el hero', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmanasdfnmgnbfsdadvb']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        const q = 'batmanasdfnmgnbfsdadvb';

        expect( wrapper.find('.alert-danger').text().trim() ).toBe(`There is no a hero with ${q}`);
        expect( wrapper ).toMatchSnapshot();

    });
    
    test('Debe llamar el push del history', () => {
        
        const historyMock = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route
                    path="/search"
                    component={ (props) => <SearchScreen history={ historyMock } /> }
                />
            </MemoryRouter>
        );

        // Hacer cambio en la caja de texto
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchHero',
                value: 'batman'
            }
        });

        // Hacer submit del formulario
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect( historyMock.push ).toHaveBeenCalledWith('?q=batman');
        
    });
    
    
});

