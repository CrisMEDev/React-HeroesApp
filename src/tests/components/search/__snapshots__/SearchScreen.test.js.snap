// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 w-100 btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrar a batman y el input con el valor de queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 w-100 btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="row g-4"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="col-md-6"
        >
          <img
            alt="Batman"
            className="img-fluid rounded-start"
            src="./assets/heroes/dc-batman.jpg"
          />
        </div>
        <div
          className="col-md-6"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Bruce Wayne
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Detective Comics #27
              </small>
            </p>
            <a
              className="btn btn-outline-dark"
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              About...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrar un mensaje de error si no se muestra el hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batmanasdfnmgnbfsdadvb"
        />
        <button
          className="btn m-1 w-100 btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a hero with 
        batmanasdfnmgnbfsdadvb
      </div>
    </div>
  </div>
</div>
`;
